src/position.h

```diff
@@ -950,9 +950,39 @@ inline bool Position::flag_move() const {
 
 inline bool Position::flag_reached(Color c) const {
   assert(var != nullptr);
-  return   (flag_region(c) & pieces(c, flag_piece(c)))
+  bool simpleResult = 
+        (flag_region(c) & pieces(c, flag_piece(c)))
         && (   popcount(flag_region(c) & pieces(c, flag_piece(c))) >= var->flagPieceCount
             || (var->flagPieceBlockedWin && !(flag_region(c) & ~pieces())));
+      
+  if (simpleResult&&var->flagPieceSafe)
+  {
+      Bitboard piecesInFlagZone = flag_region(c) & pieces(c, flag_piece(c));
+      int potentialPieces = (popcount(piecesInFlagZone));
```

Comment:
Since both attackers_to and legality checking are very expensive I would suggest to have an early return if potentialPieces >= var->flagPieceCount + 1, as already reasoned in your comment. With that also the logic becomes simpler, because once you found an attack you can immediately return false, since potentialPieces == var->flagPieceCount - 1 already is implicit by then.